### PR TITLE
add preload feature for containerd

### PR DESCRIPTION
**Pull request by medyagh**
### note1
the times in this PR should be interprteded Relative to each other, since I run it on a very old computer.
on  modern computer all of the the times bellow are -20 seconds.

### After this PR
make containerd 33 start seconds faster !

### Revolutionary finding !

containerd is 22 seconds faster than docker-runtime (tested both on docker driver see bellow)
this makes docker driver with containerd Fastest Possible local kubernetes option 
which makes sense, since our docker service binds to containerd service.
but containerd only needs 1 service (we dont need run both docker and containerd service)

to run this PR locally see the footnote bellow.

### After this PR
```
$ make && time ./out/minikube start --driver=docker --container-runtime=containerd  --kubernetes-version=v1.17.5
make: 'out/minikube' is up to date.
😄  minikube v1.9.2 on Ubuntu 20.04
✨  Using the docker driver based on user configuration
👍  Kubernetes 1.18.0 is now available. If you would like to upgrade, specify: --kubernetes-version=1.18.0
👍  Starting control plane node minikube in cluster minikube
🔥  Creating docker container (CPUs=2, Memory=3900MB) ...
📦  Preparing Kubernetes v1.17.5 on containerd 1.3.3-14-g449e9269 ...
    ▪ kubeadm.pod-network-cidr=10.244.0.0/16
🕵️   Verifying Kubernetes Components:
    🔎 verifying node conditions ...
    🔎 verifying api server ...
    🔎 verifying system pods ...
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube"

real    0m41.250s
user    0m2.728s
sys     0m0.900s
```

### Before this PR
(note that the images are cached without cached images it is even much slower)
```
$ make && time ./out/minikube start --driver=docker --container-runtime=containerd  --kubernetes-version=v1.17.5
make: 'out/minikube' is up to date.
😄  minikube v1.9.2 on Ubuntu 20.04
✨  Using the docker driver based on user configuration
👍  Kubernetes 1.18.0 is now available. If you would like to upgrade, specify: --kubernetes-version=1.18.0
👍  Starting control plane node minikube in cluster minikube
🔥  Creating docker container (CPUs=2, Memory=3900MB) ...
📦  Preparing Kubernetes v1.17.5 on containerd 1.3.3-14-g449e9269 ...
    ▪ kubeadm.pod-network-cidr=10.244.0.0/16
🕵️   Verifying Kubernetes Components:
    🔎 verifying node conditions ...
    🔎 verifying api server ...
    🔎 verifying system pods ...
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube"

real    1m13.192s
user    0m4.132s
sys     0m1.376s
```

### Docker runtime with same version of kubernetes (on my machine) is 20 seconds slower than containerd

```
$ make && time ./out/minikube start --driver=docker --container-runtime=docker  --kubernetes-version=v1.17.5

real    1m1.998s
user    0m3.634s
sys     0m1.502s
```

### Footnote1

for this PR, I generated the preload tar using 'make upload-preloaded-images-tar' and manually put it in '~/.minikube/cache/preloaded-tarball/'

once this PR is merged, the generating script should be uploading containerd tarballs as well.

### TODOs:
added an investigation that @priyawadhwa could provide good insight on this 
https://github.com/kubernetes/minikube/issues/7804

### CODE REVIEW

File: hack/preload-images/generate.go [hunk 1]
```diff
@@ -68,44 +72,77 @@ func generateTarball(kubernetesVersion, containerRuntime, tarballFilename string
...
+	sv, err := util.ParseKubernetesVersion(constants.DefaultKubernetesVersion)
+	if err != nil {
+		return errors.Wrap(err, "Failed to parse kubernetes version")
+	}
+
+	co := cruntime.Config{
+		Type:              containerRuntime,
+		Runner:            runner,
+		ImageRepository:   "",
+		KubernetesVersion: sv, // I think this is just to statsify cruntime and shouldnt matter
```
Comment (priyawadhwa) [inline hack/preload-images/generate.go:91]: nit: I'd remove "I think" and check the spelling of satisfy

File: hack/preload-images/generate.go [hunk 2]
```diff
@@ -68,44 +72,77 @@ func generateTarball(kubernetesVersion, containerRuntime, tarballFilename string
...
+	if err != nil {
+		exit.WithError("Failed runtime", err)
+	}
+	if err := cr.Enable(true); err != nil {
+		exit.WithError("enable container runtime ", err)
+	}
+
 	for _, img := range imgs {
 		cmd := exec.Command("docker", "exec", profile, "docker", "pull", img)
+		if containerRuntime == "containerd" {
```
Comment (priyawadhwa) [inline hack/preload-images/generate.go:103]: Could we move this into a helper function? Something like:

```go
func imagePullCommand(containerRuntime, img string) *exec.Cmd {
}
```

File: pkg/drivers/kic/kic.go [hunk 3]
```diff
@@ -130,7 +130,7 @@ func (d *Driver) Create() error {
 			return
 		}
 		t := time.Now()
-		glog.Infof("Starting extracting preloaded images to volume")
+		glog.Infof("Starting extracting preloaded images to volume ....")
```
Comment (priyawadhwa) [inline pkg/drivers/kic/kic.go:133]: nit: we typically use three dots for logs

File: pkg/minikube/cruntime/containerd.go [hunk 4]
```diff
@@ -310,5 +314,110 @@ func (r *Containerd) Preload(cfg config.KubernetesConfig) error {
...
+	}
+
+	// Make sure images == imgs
+	for _, i := range images {
+		found := false
+		for _, ji := range jsonImages.Images {
+			for _, rt := range ji.RepoTags {
+				// in crictl docker.io is appened to images without repo organization
+				// for example "kubernetesui/dashboard:v2.0.0-rc6 will show up as "docker.io/kubernetesui/dashboard:v2.0.0-rc6"
+				if !strings.Contains(i, ".io/") {
```
Comment (priyawadhwa) [inline pkg/minikube/cruntime/containerd.go:403]: It would be nice to have a unit test for this

File: pkg/minikube/machine/cache_images.go [hunk 5]
```diff
@@ -64,12 +64,20 @@ func CacheImagesForBootstrapper(imageRepository string, version string, clusterB
 
 // LoadImages loads previously cached images into the container runtime
 func LoadImages(cc *config.ClusterConfig, runner command.Runner, images []string, cacheDir string) error {
-	// Skip loading images if images already exist
-	if cruntime.DockerImagesPreloaded(runner, images) {
-		glog.Infof("Images are preloaded, skipping loading")
-		return nil
+	if cc.KubernetesConfig.ContainerRuntime == "docker" {
```
Comment (priyawadhwa) [inline pkg/minikube/machine/cache_images.go:67]: We could move these into a helper function as well that could be called throughout the code:

```go
func ImagesPreloaded(containerRuntime string) bool }
}
```

Comment (medyagh) [reply pkg/minikube/machine/cache_images.go:67]: good idea

File: pkg/minikube/cruntime/containerd.go (hunk 4 above)
Comment (medyagh) [reply pkg/minikube/cruntime/containerd.go:403]: good idea !

File: pkg/drivers/kic/kic.go (hunk 3 above)
Comment (medyagh) [reply pkg/drivers/kic/kic.go:133]: done

File: hack/preload-images/generate.go (hunk 2 above)
Comment (medyagh) [reply hack/preload-images/generate.go:103]: done

File: pkg/minikube/machine/cache_images.go (hunk 5 above)
Comment (medyagh) [reply pkg/minikube/machine/cache_images.go:67]: done

File: pkg/minikube/cruntime/containerd _test.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (priyawadhwa) [inline pkg/minikube/cruntime/containerd _test.go:1]: This filename has an extra space in it, looks like it's called `containerd _test.go`

File: pkg/minikube/cruntime/containerd.go [hunk 7]
```diff
@@ -310,5 +314,118 @@ func (r *Containerd) Preload(cfg config.KubernetesConfig) error {
...
+
+	k8sVersion := cfg.KubernetesVersion
+	cRuntime := cfg.ContainerRuntime
+
+	// If images already exist, return
+	images, err := images.Kubeadm(cfg.ImageRepository, k8sVersion)
+	if err != nil {
+		return errors.Wrap(err, "getting images")
+	}
+	if containerdImagesPreloaded(r.Runner, images) {
```
Comment (priyawadhwa) [inline pkg/minikube/cruntime/containerd.go:326]: we can use `ImagesPreloaded` here now

File: pkg/minikube/cruntime/containerd _test.go (hunk 6 above)
Comment (medyagh) [reply pkg/minikube/cruntime/containerd _test.go:1]: done

File: pkg/minikube/cruntime/containerd.go (hunk 7 above)
Comment (medyagh) [reply pkg/minikube/cruntime/containerd.go:326]: done

File: pkg/minikube/cruntime/containerd _test.go (hunk 6 above)
Comment (medyagh) [reply pkg/minikube/cruntime/containerd _test.go]: done

**Comment by medyagh** [comment]
crio test failures is related to https://github.com/kubernetes/minikube/issues/7808
https://storage.googleapis.com/minikube-builds/logs/7793/49fa90a/Docker_Linux.html#fail_TestStartStop%2fgroup%2fcrio

**Pull request merged**